File: app/src/main/java/com/examle/curexchange/di/module/AppModule.java
package com.examle.curexchange.di.module;

import android.arch.persistence.room.Room;
import android.content.Context;

import com.examle.curexchange.App;
import com.examle.curexchange.Constants;
import com.examle.curexchange.data.database.AppDatabase;
import com.examle.curexchange.data.database.dao.CurrencyDao;
import com.examle.curexchange.data.database.dao.HistoryDao;
import com.examle.curexchange.data.remote.ApiCryptoCode;
import com.examle.curexchange.data.remote.ApiExchange;

import javax.inject.Named;
import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

@Module
public class AppModule {

    private final Context context;

    public AppModule(App context) {
        this.context = context;
    }

    @Provides
    @Singleton
    Context provideContext() {
        return context;
    }

    @Provides
    @Singleton
    App provideApp() {
        return App.getApp();
    }

    @Provides
    @Singleton
    @Named("codes")
    Retrofit provideRetrofitCryptoCode() {
        return new Retrofit.Builder()
                .baseUrl(Constants.CRYPTO_CODE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();
    }

    @Provides
    @Singleton
    ApiCryptoCode provideCryptoCodeApi(@Named("codes") Retrofit retrofit) {
        return retrofit.create(ApiCryptoCode.class);
    }

    @Provides
    @Singleton
    @Named("exchange")
    Retrofit provideExchangeRetrofit() {
        return new Retrofit.Builder()
                .baseUrl(Constants.EXCHANGE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();
    }

    @Provides
    @Singleton
    ApiExchange provideExchangeApi(@Named("exchange") Retrofit retrofit) {
        return retrofit.create(ApiExchange.class);
    }

    @Provides
    @Singleton
    AppDatabase provideAppDatabase(App app) {
        return Room
                .databaseBuilder(app, AppDatabase.class, Constants.DATABASE_NAME)
                .build();
    }

    @Provides
    @Singleton
    CurrencyDao provideCurrencyDao(AppDatabase appDatabase) {
        return appDatabase.currencyDao();
    }

    @Provides
    @Singleton
    HistoryDao provideHistoryDao(AppDatabase appDatabase) {
        return appDatabase.historyDao();
    }
}

File: app/src/main/java/com/examle/curexchange/di/module/AppModule.kt
package com.examle.curexchange.di.module

import android.arch.persistence.room.Room
import com.examle.curexchange.App
import com.examle.curexchange.Constants
import com.examle.curexchange.data.database.AppDatabase
import com.examle.curexchange.data.database.dao.CurrencyDao
import com.examle.curexchange.data.database.dao.HistoryDao
import com.examle.curexchange.data.remote.ApiCryptoCode
import com.examle.curexchange.data.remote.ApiExchange
import dagger.Module
import dagger.Provides
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import javax.inject.Named
import javax.inject.Singleton


@Module
class AppModule(val app: App) {

    @Provides
    @Singleton
    fun provideApp() = app

    @Provides
    @Singleton
    @Named("codes")
    fun provideRetrofitCryptoCode(): Retrofit {
        return Retrofit.Builder()
                .baseUrl(Constants.CRYPTO_CODE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build()
    }

    @Provides
    @Singleton
    fun provideCryptoCodeApi(@Named("codes") retrofit: Retrofit): ApiCryptoCode {
        return retrofit.create(ApiCryptoCode::class.java)
    }

    @Provides
    @Singleton
    @Named("exchange")
    fun provideExchangeRetrofit(): Retrofit {
        return Retrofit.Builder()
                .baseUrl(Constants.EXCHANGE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build()
    }

    @Provides
    @Singleton
    fun provideExchangeApi(@Named("exchange") retrofit: Retrofit): ApiExchange {
        return retrofit.create(ApiExchange::class.java)
    }

    @Provides
    @Singleton
    fun provideAppDatabase(app: App): AppDatabase {
        return Room
                .databaseBuilder(app, AppDatabase::class.java, Constants.DATABASE_NAME)
                .build()
    }

    @Provides
    @Singleton
    fun provideCurrencyDao(appDatabase: AppDatabase): CurrencyDao {
        return appDatabase.currencyDao()
    }

    @Provides
    @Singleton
    fun provideHistoryDao(appDatabase: AppDatabase): HistoryDao {
        return appDatabase.historyDao()
    }


}